### PR TITLE
Support treestyletab

### PR DESCRIPTION
**Pull request by piroor**
This change will help Tree Style Tab's user including me. How about that?

(This PR depends on #4.)

### CODE REVIEW

File: src/js/background.js [hunk 1]
```diff
@@ -225,6 +225,7 @@
           url, windowId,
           active: true,
           index: tabIndex + 1,
+          openerTabId: data.tab.id,
```
Comment (asamuzaK) [inline src/js/background.js]: openerTabId: tabId,

**Comment by asamuzaK** [comment]
Before implementing this PR, I think that it is necessary to ask users whether to allow cooperating with other extensions.
I will implement the opt-in for that, so can you wait until then?

**Comment by asamuzaK** [comment]
I've implemented external extension handler in #6 
Would you rebase and commit?

**Comment by asamuzaK** [comment]
> I will implement the opt-in for that, so can you wait until then?

For note.
`management` could not be optional_permissions in Firefox.
So, I added `management` to default permissions which will be asked when install.

**Comment by piroor** [comment]
Thanks. I've updated this PR with the new framework.

File: src/js/background.js [hunk 2]
```diff
@@ -103,14 +103,21 @@
     return func || null;
   };
 
+  const registerToExtTST = async () => sendMsg(EXT_TST, {
```
Comment (piroor) [inline src/js/background.js]: This looks not good - this should be defined somewhere under a namespace just for the external addon. Any ideas?

Comment (asamuzaK) [inline src/js/background.js]: JSDOC please.
And remove `browser.` namespace.

File: src/js/background.js [hunk 3]
```diff
@@ -103,14 +103,21 @@
...
+    type: "register-self",
+    name: browser.i18n.getMessage("extensionName"),
+    icons: browser.runtime.getManifest().icons,
```
Comment (asamuzaK) [inline src/js/background.js]: Isn't runtime.getManifest() async?
Maybe it's better to await for icons and sendMsg()
Something like:
```
const registerToExtTST = async () => {
  const {icons} = await runtime.getManifest();
  return sendMsg(EXT_TST, {
    icons,
    ...
  });
};
```

File: src/js/background.js [hunk 4]
```diff
@@ -386,7 +398,18 @@
     const func = [];
     if (id && externalExts.has(id)) {
       if (id === EXT_TST) {
-        // handle message from tst
+        switch (msg.type) {
+          case "ready": {
+            func.push(setExternalExts().then(registerToExtTST)
```
Comment (asamuzaK) [inline src/js/background.js]: Nit, please break line after `push(` opener paren.
```
func.push(
  setExternalExts().then().then()
);
```

**Review by asamuzaK — CHANGES_REQUESTED** [review]

File: src/js/background.js (hunk 3 above)
Comment (piroor) [reply src/js/background.js]: The [API document](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/getManifest) says that it returns a JSON object synchronously.

**Comment by piroor** [comment]
Thanks, I've fixed these mistakes.

File: src/js/background.js (hunk 3 above)
Comment (asamuzaK) [reply src/js/background.js]: I didn't know that, thanks.

File: src/js/background.js (hunk 2 above)
Comment (asamuzaK) [reply src/js/background.js]: You should move `sendMsg()` inside `handleExternalExtsRequirements` and on "ready" message call `handleExternalExtsRequirements` again?

Maybe in the future, even when handling other extensions requirements, a register message will be passed to TST, though.

**Comment by asamuzaK** [comment]
Other idea for registerToExtTST.

Swith `externalExts` to Map();

In handleExternalExtsRequirements(), register function.
```
if (externalExts.has(EXT_TST)) {
  const register = () => sendMsg(EXT_TST, {
    ...
  });
  externalExts.set(EXT_TST, {
    register,
  });
  func.push(register());
});
```

and in handleMsg(),
```
if (id === EXT_TST) {
  switch (msg.type) {
    case "ready": {
      if (externalExts.has(EXT_TST)) {
        const {register} = externalExts.get(EXT_TST);
        if (typeof register === "function") {
          func.push(
            register(),
            createMenu(),
          );
        }
        ...
      } else {
        func.push(setExternalExts().then(...));
      }
      break;
    }
    case ...
  }
}
```

How that sound?

**Review by asamuzaK — APPROVED** [review]

File: src/js/background.js (hunk 2 above)
Comment (piroor) [reply src/js/background.js]: Sorry, I misunderstood as: `management.getAll()` don't return the information of TST until TST is loaded (even if it is installed). After some expreriments I've realized that it returns all installed extensions and `externalExts` always have TST's information even if TST is initialized after shareSNS. You're right, now I'm changing code to reuse `handleExternalExtsRequirements` itself.

**Comment by piroor** [comment]
I think that the commit f83efba should solve the concern told at https://github.com/asamuzaK/shareSNS/pull/5#discussion_r205730718

**Review by asamuzaK — APPROVED** [review]

**Comment by asamuzaK** [comment]
Thanks a lot!
I found several things to do about extenal extension handling, I'll take care of the rest.

**Comment by asamuzaK** [comment]
Try build [shareSNS/dest at master · asamuzaK/shareSNS](https://github.com/asamuzaK/shareSNS/tree/master/dest "shareSNS/dest at master · asamuzaK/shareSNS")

**Pull request merged**